transport/http/outbound.go

```diff
@@ -256,6 +256,13 @@ func (o *Outbound) call(ctx context.Context, treq *transport.Request) (*transpor
 
 	span.SetTag("http.status_code", response.StatusCode)
 
+	// Service name match validation, return yarpcerrors.CodeInternal error if not match
+	if match, resSvcName := checkServiceMatch(treq.Service, response.Header); !match {
+		return nil, transport.UpdateSpanWithErr(span,
+			yarpcerrors.InternalErrorf("service name sent from the request "+
+				"does not match the service name received in the response, sent %q, got: %q", treq.Service, resSvcName))
+	}
+
```

Comment:
Might we want a hook to allow emitting metrics or logs in the case of a permitted empty service header response (or similar UpdateSpanWithErr on empty service header in response if strict enforcement is desired by the caller)?

Reply:
thanks for the advice, could you explain the idea a little bit? I'm not quite get yet, did you mean we need to append empty header event into tracing span?

Reply:
Similar to my response above, this is probably unnecessary. Older YARPC libraries calling new libraries would be affected if we enabled strict enforcement and that may just end up being noise.

Reply:
I think I didn't communicate my suggestion clearly enough. I am suggesting that it remain essentially transparent now, but coded in such a way that a user can choose to ratchet up enforcement easily if/when desired. There are 3 states to consider:
- response missing service name header (`if match && resSvcName == "")
- response with non-matching service name header (`if !match`)
- response with matching service name header (`if match && resSvcName != "")

We have complete enough information to handle all three differently. I think we have an opportunity in the first condition to get visibility into services not responding with the header and we should take it. I think it should be distinguished from an actual match, if only by debug logging today. Some users might want/need to set enforcement of this condition as an error sooner than others. We should wire that up now, but leave it permissive by default.

nit: "match" is a misnomer in the case of no header in response since there was no header to match against

Reply:
@zmt That makes sense. I believe the original request for this feature did include an ask for an eventual "strict" mode for clients. Although, rather than providing a user-facing hook for the first (or any of the cases), my preference would be to just log that internally for now. All transports will have access to a logger once #1479 lands.

This lets us defer how the API for invoking strict mode looks at the client-level.